pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,121 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	// dockerIgnoreFile defines the Name of the scanner file
+	dockerIgnoreFile = ".dockerignore"
+	// invertPrefix is used to prefix the Patterns to symbolize inversions
+	invertPrefix = "!"
+)
+
+var (
+	_ FileIgnorer = &defaultIgnorer{}
+	// patterns holds all the possible Ignorable patterns
+	ignorePatterns []fileIgnorePattern
+)
+
+// dockerIgnorer processes .dockerignore and use the extracted Patterns for ignoring files
+type dockerIgnorer struct {
+	// root directory to start the scanner
+	directory string
+	// index of all excludes directories
+	excludedDirs sets.String
+}
+
+// NewOrDefault builds and returns the new or default FileIgnorer interface
+// In all file not found cases the this wil return the default ignorer
+func NewOrDefault(dir string) (FileIgnorer, error) {
+	var rawPatterns []string
+	// every new gets clean includes
+	ignorePatterns = make([]fileIgnorePattern, 0)
+	ignoreFile := filepath.Join(dir, dockerIgnoreFile)
+
+	err := scanAndBuildPatternsList(ignoreFile, &rawPatterns)
+
+	dftIgnorer := &defaultIgnorer{
+		directory: dir,
+	}
+
+	if err != nil {
+		return dftIgnorer, nil
+	}
+
+	for _, ip := range rawPatterns {
+		x := toFileIgnorePattern(dir, ip)
+		ignorePatterns = append(ignorePatterns, *x)
+	}
+
+	return &dockerIgnorer{directory: dir, excludedDirs: sets.NewString(defaultPatterns...)}, nil
+}
+
+// CanIgnore is the directory scanner to scan directory for .dockerignore based
+// ignore patterns
+func (i *dockerIgnorer) CanIgnore(path string, fi os.FileInfo) (Ignorable, error) {
+	return scanDir(i.directory, path, fi, i.excludedDirs)
+}
+
+// scanAndBuildPatternsList takes the file typically the .dockerignore file
+// splits the file by new line (\n) and normalize them with following rules
+// - removes the UTF8 Byte Order Mark (BOM) characters
+// - scanner comments line starting with "#"
+// - trim spaces in the pattern
+// - make the pattern as clean filenames using golang filepath utils
+func scanAndBuildPatternsList(ignoreFile string, patterns *[]string) error {
+	*patterns = append(*patterns, defaultPatterns...)
+
+	fr, err := os.Open(ignoreFile)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(fr)
+
+	// UTF8 byte order mark (BOM) which are typically first three bytes of the file with
+	// the hexadecimal characters: EF,BB,BF
+	utf8bom := []byte{0xEF, 0xBB, 0xBF}
+	lineNo := 0
+
+	for scanner.Scan() {
+		scannedBytes := scanner.Bytes()
+
+		if lineNo == 0 {
+			scannedBytes = bytes.TrimPrefix(scannedBytes, utf8bom)
+		}
+		lineNo++
+		p := sanitizePattern(string(scannedBytes))
+		if p != "" {
```

Comment:
```suggestion
		if p := sanitizePattern(string(scannedBytes)); p != "" {
```

Reply:
https://github.com/mattmoor/mink/pull/308#discussion_r542453492

I don't understand why need it ? since the NewOrDefault covers both the flows ?? Am I missing something 🤔

Reply:
I moved the closing of the filereader(`fr`) inside of the scanAndBuildPatterns method

https://github.com/mattmoor/mink/pull/308/files#diff-982056932008369254af0223a5f43c379b9566220dda956d3f7be9067348d975R115